api/buckets/service.go

```diff
@@ -1584,6 +1809,155 @@ func (s *Service) removeNodeAtPath(ctx context.Context, pth path.Path, key []byt
 	return path.IpfsPath(np[0].new.Cid()), nil
 }
 
+func (s *Service) PushPathAccessRoles(ctx context.Context, req *pb.PushPathAccessRolesRequest) (*pb.PushPathAccessRolesResponse, error) {
+	log.Debugf("received push path access roles request")
+
+	dbID, ok := common.ThreadIDFromContext(ctx)
+	if !ok {
+		return nil, fmt.Errorf("db required")
+	}
+	dbToken, _ := thread.TokenFromContext(ctx)
+
+	reqPath, err := parsePath(req.Path)
+	if err != nil {
+		return nil, err
+	}
+	buck, pth, err := s.getBucketPath(ctx, dbID, req.Key, reqPath, dbToken)
+	if err != nil {
+		return nil, err
+	}
+	linkKey := buck.GetLinkEncryptionKey()
+	pathNode, err := s.getNodeAtPath(ctx, pth, linkKey)
+	if err != nil {
+		return nil, err
+	}
+
+	var currentFileKeys map[string][]byte
+	md, mdPath, ok := buck.GetMetadataForPath(reqPath, false)
+	if !ok {
+		return nil, fmt.Errorf("could not resolve path: %s", reqPath)
+	}
+	var target tdb.Metadata
+	if mdPath != reqPath { // If the metadata is inherited from a parent, create a new entry
+		target = tdb.Metadata{
+			Roles: make(map[string]buckets.Role),
+		}
+	} else {
+		target = md
+	}
+	if buck.IsPrivate() {
+		currentFileKeys, err = buck.GetFileEncryptionKeysForPrefix(reqPath)
+		if err != nil {
+			return nil, err
+		}
+		newFileKey, err := dcrypto.NewKey()
+		if err != nil {
+			return nil, err
+		}
+		target.SetFileEncryptionKey(newFileKey) // Create or refresh the file key
+	}
+
+	newRoles, err := buckets.RolesFromPb(req.Roles)
+	if err != nil {
+		return nil, err
+	}
+	var changed bool
+	for k, r := range newRoles {
+		if x, ok := target.Roles[k]; ok && x == r {
+			continue
+		}
+		if r > 0 {
+			target.Roles[k] = r
+		} else {
+			delete(target.Roles, k)
+		}
+		changed = true
+	}
+	if changed {
+		buck.UpdatedAt = time.Now().UnixNano()
+		target.UpdatedAt = buck.UpdatedAt
+		buck.Metadata[reqPath] = target
+
+		txn, err := s.Buckets.WriteTxn(ctx, dbID, tdb.WithToken(dbToken))
+		if err != nil {
+			return nil, err
+		}
+		if err = txn.Verify(ctx, buck); err != nil {
+			return nil, err
+		}
+
+		if buck.IsPrivate() {
+			if err := buck.RotateFileEncryptionKeysForPrefix(reqPath); err != nil {
+				return nil, err
+			}
+			newFileKeys, err := buck.GetFileEncryptionKeysForPrefix(reqPath)
+			if err != nil {
+				return nil, err
+			}
+			nmap, err := s.encryptDag(ctx, s.IPFSClient.Dag(), pathNode, reqPath, linkKey, currentFileKeys, newFileKeys, nil, nil)
+			if err != nil {
+				return nil, err
+			}
+			nodes := make([]ipld.Node, len(nmap))
+			i := 0
+			for _, tn := range nmap {
+				nodes[i] = tn.node
+				i++
+			}
+			pn := nmap[pathNode.Cid()].node
+			dirPath, err := s.insertNodeAtPath(ctx, pn, path.Join(path.New(buck.Path), reqPath), linkKey)
+			if err != nil {
+				return nil, fmt.Errorf("updating pinned root: %s", err)
+			}
+			if err := s.addAndPinNodes(ctx, nodes); err != nil {
+				return nil, err
+			}
+			buck.Path = dirPath.String()
+		}
```

Comment:
Q: So the old subtree was unpined as a result of above methods, right?

Reply:
Yep, `insertNodeAtPath` handles unpinning old branches either directly or by a pin replace.